src/main/scala/viper/gobra/translator/util/Core.scala

```diff
@@ -0,0 +1,109 @@
+//
+// Copyright (c) 2011-2020 ETH Zurich.
+
+package viper.gobra.translator.util
+
+import viper.gobra.ast.{internal => in}
+
+import scala.annotation.unused
+
+object Core {
+  def unit[R](x: R): Core[R] = Core { args => (x, args) }
+
+  def vector[R](x: Vector[Core[R]]): Core[Vector[R]] =
+    x.foldRight(unit(Vector.empty[R])){ case (l, r)  => for (x <- l; xs <- r) yield x +: xs }
+
+  def option[R](x: Option[Core[R]]): Core[Option[R]] = x.fold[Core[Option[R]]](unit(None))(_.map(Some(_)))
+
+  /**
+    * For a statements s, the result is a function from variables to s where all sub-expressions are substituted with the argument variables.
+    * The result satisfies `core(s).run(coreArgs(s)) == s`
+    * */
+  def core(x: in.Deferrable): Core[in.Deferrable] = x match {
+    case x: in.FunctionCall => for (args <- vector(x.args map core)) yield x.copy(args = args)(x.info)
+    case x: in.MethodCall => for (recv <- core(x.recv); args <- vector(x.args map core)) yield x.copy(recv = recv, args = args)(x.info)
+    case x: in.Fold => for (acc <- core(x.acc)) yield x.copy(acc = acc)(x.info)
+    case x: in.Unfold => for (acc <- core(x.acc)) yield x.copy(acc = acc)(x.info)
+    case x: in.PredExprFold => for (base <- deepCore(x.base); args <- vector(x.args map core)) yield x.copy(base = base, args = args)(x.info)
+    case x: in.PredExprUnfold => for (base <- deepCore(x.base); args <- vector(x.args map core)) yield x.copy(base = base, args = args)(x.info)
+  }
+
+  def core(x: in.Access): Core[in.Access] =
+    for (e <- core(x.e); p <- core(x.p)) yield x.copy(e = e, p = p)(x.info)
+
+  def core(x: in.Accessible): Core[in.Accessible] = x match {
+    case x: in.Accessible.Predicate => for (op <- core(x.op)) yield x.copy(op = op)
+    case x: in.Accessible.ExprAccess => for (op <- core(x.op)) yield x.copy(op = op)
+    case x: in.Accessible.Address => for (op <- core(x.op)) yield x.copy(op = op)
+    case x: in.Accessible.PredExpr => for (op <- core(x.op)) yield x.copy(op = op)
+  }
+
+  def core(x: in.PredicateAccess): Core[in.PredicateAccess] = x match {
+    case x: in.FPredicateAccess =>
+      for (args <- vector(x.args map core)) yield x.copy(args = args)(x.info)
+    case x: in.MPredicateAccess =>
+      for (recv <- core(x.recv); args <- vector(x.args map core)) yield x.copy(args = args, recv = recv)(x.info)
+    case x: in.MemoryPredicateAccess =>
+      for (arg <- core(x.arg)) yield x.copy(arg = arg)(x.info)
+  }
+
+  def core(x: in.PredExprInstance): Core[in.PredExprInstance] = {
+    for (base <- core(x.base); args <- vector(x.args map core)) yield x.copy(base = base, args = args)(x.info)
+  }
+
+  def core(@unused x: in.Expr): Core[in.LocalVar] = Core { args => (args.head, args.tail) }
+
+  def deepCore(x: in.PredicateConstructor): Core[in.PredicateConstructor] = {
+    for {
+      args <- vector(x.args.map(y => option(y.map(core))))
+    } yield x.copy(args = args)(x.info)
+  }
+
```

Comment:
Is `deepCore` meant to be accessible from the outside, or is it an internal auxiliar method which we can make private?